src/AzureOpenAIProxy.Management/Services/MetricService.cs

```diff
@@ -0,0 +1,272 @@
+            EventId = eventId,
+            AttendeeCount = attendeeCount,
+            RequestCount = requestCount,
+            ModelData = modeldata
+        };
+    }
+
+    private async Task<ModelData> GetModelCountAsync(string eventId)
+    {
+        if (conn.State != ConnectionState.Open)
+            await conn.OpenAsync();
+
+        using var modelCountCommand = conn.CreateCommand();
+        modelCountCommand.CommandText = """
+        SELECT event_id, date_stamp, resource, SUM(prompt_tokens) AS prompt_tokens, SUM(completion_tokens) AS completion_tokens, SUM(total_tokens) AS total_tokens, COUNT(*) AS requests
+        FROM aoai.metric_view where event_id = @EventId
+        GROUP BY date_stamp, event_id, resource
+        ORDER BY requests DESC
+        """;
+
+        modelCountCommand.Parameters.Add(new NpgsqlParameter("EventId", eventId));
+        using var reader = await modelCountCommand.ExecuteReaderAsync();
+
+        var metricsData = new List<MetricsData>();
+        while (reader.Read())
+        {
+            var item = new MetricsData
+            {
+                EventId = eventId,
+                DateStamp = reader.GetDateTime(1),
+                Resource = reader.IsDBNull(2) ? "Unknown" : reader.GetString(2),
+                PromptTokens = reader.IsDBNull(3) ? 0 : reader.GetInt64(3),
+                CompletionTokens = reader.IsDBNull(4) ? 0 : reader.GetInt64(4),
+                TotalTokens = reader.IsDBNull(5) ? 0 : reader.GetInt64(5),
+                Requests = reader.IsDBNull(6) ? 0 : reader.GetInt64(6)
+            };
+
+            metricsData.Add(item);
+        };
+
+        var summary = metricsData
+            .GroupBy(r => new { EventId = r.EventId, Resource = r.Resource })
+            .Select(g => new
+            {
+                g.Key.Resource,
+                PromptTokens = g.Sum(x => (long)x.PromptTokens),
+                CompletionTokens = g.Sum(x => (long)x.CompletionTokens),
+                TotalTokens = g.Sum(x => (long)x.TotalTokens),
+                Requests = g.Sum(x => (long)x.Requests)
+            })
+            .OrderByDescending(x => x.Requests);
+
+        List<ModelCounts> modelCounts = summary.Select(item => new ModelCounts
+        {
+            Resource = item.Resource,
+            Count = (int)item.Requests,
+            PromptTokens = item.PromptTokens,
+            CompletionTokens = item.CompletionTokens,
+            TotalTokens = item.TotalTokens
+        }).ToList();
+
+        // Create Line Chart Data X axis is DateStamp and Y axis is Requests
+        List<ChartData> chartData = metricsData
+            .GroupBy(r => r.DateStamp)
+            .Select(g => new ChartData
+            {
+                DateStamp = g.Key,
+                Count = g.Sum(x => x.Requests)
+            })
+            .OrderBy(x => x.DateStamp)
+            .ToList();
+
+        long runningTotal = 0;
+        chartData.ForEach(x => runningTotal = x.Count += runningTotal);
```

Comment:
Don't use the `ForEach` extension, use [Enumerable.Aggregate](https://learn.microsoft.com/en-us/dotnet/api/system.linq.enumerable.aggregate?view=net-8.0). Also, this would save the `ToList` call on line 155, resulting in better lazy evaluation.